packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -0,0 +1,624 @@
+        if (x >= node.rect.left && x <= node.rect.right) {
+          return node;
+        }
+        if (x < node.rect.left) {
+          max = mid;
+        }
+        if (x > node.rect.right) {
+          min = mid + 1;
+        }
+      }
+      return null;
+    }
+
+    return nodes.isEmpty ? null : binarySearchForNode();
+  }
+
+  int getRowIndexForY(double y) {
+    if (y < flameChartTop) {
+      return -1;
+    }
+    return math.max((y - flameChartTop) ~/ rowHeightWithPadding, 0);
+  }
+}
+
+class FlameChartCanvas extends FlameChart {
+  FlameChartCanvas({
+    @required CpuProfileData data,
+    @required flameChartWidth,
+    @required flameChartHeight,
+  }) : super(
+          data: data,
+          flameChartWidth: flameChartWidth,
+          flameChartHeight: flameChartHeight,
+        ) {
+    _viewportCanvas = ViewportCanvas(
+      paintCallback: _paintCallback,
+      onTap: _onTap,
+      classes: 'ui-details-section cpu-flame-chart',
+    )..element.element.style.overflow = 'hidden';
+
+    _viewportCanvas.setContentSize(flameChartWidth, flameChartHeight);
+
+    _dragScroll.enableDragScrolling(element);
+    _dragScroll.onVerticalScroll = () {
+      _viewportCanvas.rebuild(force: true);
+    };
+
+    element.element.onMouseWheel.listen(_handleMouseWheel);
+  }
+
+  ViewportCanvas _viewportCanvas;
+
+  CoreElement get element => _viewportCanvas.element;
+
+  /// Maximum scroll delta allowed for scrollwheel based zooming.
+  ///
+  /// This isn't really needed but is a reasonable for safety in case we
+  /// aren't handling some mouse based scroll wheel behavior well, etc.
+  final num maxScrollWheelDelta = 20;
+
+  /// Maximum zoom level we should allow.
+  ///
+  /// Arbitrary large number to accommodate spacing for some of the shortest
+  /// events when zoomed in to [_maxZoomLevel].
+  final _maxZoomLevel = 150;
+  final _minZoomLevel = 1;
+  num zoomLevel = 1;
+
+  num get _zoomMultiplier => zoomLevel * 0.003;
+
+  // The DOM doesn't allow floating point scroll offsets so we track a
+  // theoretical floating point scroll offset corresponding to the current
+  // scroll offset to reduce floating point error when zooming.
+  num floatingPointScrollLeft = 0;
+
+  // TODO(kenzie): optimize painting to canvas by grouping paints with the same
+  // canvas settings.
+  void _paintCallback(CanvasRenderingContext2D canvas, Rect rect) {
+    final int startRow = math.max(getRowIndexForY(rect.top), 0);
+    final int endRow = math.min(
+      getRowIndexForY(rect.bottom) + 1,
+      rows.length - 1,
+    );
+    for (int i = startRow; i < endRow; i++) {
+      paintRow(canvas, i, rect);
+    }
+
+    paintTimelineGrid(canvas, rect);
+  }
+
+  void paintTimelineGrid(CanvasRenderingContext2D canvas, Rect visible) {
+    timelineGrid.paint(canvas, _viewportCanvas.viewport, visible);
+  }
+
+  void paintRow(
+    CanvasRenderingContext2D canvas,
+    int index,
+    Rect visible,
+  ) {
+    canvas.save();
```

Comment:
remove this save and restore. I doubt it is needed.

Reply:
done.